File: 22.07.19(1513).cpp
#include<iostream>
#include<string.h>
#include<math.h>
#include<vector>
using namespace std;
int dp[51][51][51][51]; //y,x,count,orak
int n, m, c;
int vec[51][51];

int solve(int y, int x, int count, int  orak) {
	if (y < 1 || x < 1 || count < 0)return 0;
	int& r = dp[y][x][count][orak];
	if (r != -1)return r;
	r = 0;
	if(vec[y][x] == 0) {
		//오락실이 아닌 경우(이동 - 우측, 하단)
		r = solve(y - 1, x, count, orak) % 1000007  + solve(y, x - 1, count, orak) % 1000007;
	}
	
	//오락실인 경우 == (y,x)
	else if (vec[y][x] == orak) {
		for (int i = 0; i < orak; i++) {
			r += solve(y - 1, x, count -1 , i) % 1000007  + solve(y, x - 1, count - 1, i) % 1000007;
		}
	}

	return r % 1000007;
}


int main() {
	cin >> n >> m >> c;
	memset(vec, 0, sizeof(vec));
	int a, b;
	for (int i = 1; i <= c; i++) {
		cin >> a >> b;
		vec[a][b] = i;
		//vec.push_back({ a,b });
	}
	memset(dp, -1, sizeof(dp));
	if (vec[1][1] == 0) {
		dp[1][1][0][0] = 1;
	}
	else {
		dp[1][1][1][vec[1][1]] = 1;
	}
	int answer = 0;
	for (int i = 0; i <= c; i++) {
		answer = 0;
		for (int j = 0; j <= c; j++) {
			answer += solve(n, m, i, j)%1000007;
		}
		cout << answer%1000007 << " ";
	}
}

File: 22.07.19(1513).java
import java.awt.desktop.PreferencesEvent;
import java.awt.image.AreaAveragingScaleFilter;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.time.chrono.ThaiBuddhistEra;
import java.util.Arrays;
import java.util.Scanner;
import java.util.StringTokenizer;
import java.util.Vector;


public class Main {
    static int n,m,c;

    static int [][][][]dp = new int[51][51][51][51];
    static int [][] orak = new int[51][51];
    public static int solve(int y, int x, int count, int num){
        if(y<1 || x < 1|| count < 0)return 0;
        if(dp[y][x][count][num] != -1) return  dp[y][x][count][num];
        dp[y][x][count][num] = 0;
        if(orak[y][x] == 0){
            dp[y][x][count][num] = solve(y-1,x,count,num)%1000007 + solve(y,x-1, count,num)%1000007;
        }
        else if(orak[y][x] == num) {
            for (int i = 0; i < num; i++) {
                dp[y][x][count][num] += solve(y - 1, x, count - 1, i) % 1000007 + solve(y, x - 1, count - 1, i) % 1000007;
            }
        }
        return dp[y][x][count][num] %1000007;
    }


    public static void main(String args[]) throws IOException {
        Scanner sc = new Scanner(System.in);
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        n = sc.nextInt();
        m = sc.nextInt();
        c = sc.nextInt();
        for(int i=0; i<51; i++){
            for(int j=0; j<51; j++){
                orak[i][j]= 0;
            }
        }
        int a,b;
        for(int i=1; i<=c; i++){
            a=sc.nextInt();
            b=sc.nextInt();
            orak[a][b] = i;
        }
        for(int i=0; i<51; i++){
            for(int j=0; j<51; j++){
                for(int x=0; x<51; x++){
                    for(int y=0; y<51; y++){
                        dp[i][j][x][y]=-1;
                    }
                }
            }
        }
        if(orak[1][1] ==0){
            dp[1][1][0][0] = 1;
        }
        else{
            dp[1][1][1][orak[1][1]] = 1;
        }
        for(int i=0; i<=c; i++){
            int answer = 0;
            for(int j=0; j<=c; j++){
                answer += solve(n,m,i,j) %1000007;
            }
            System.out.print(answer%1000007 + " ");
        }
    }


}
